telemetry/metrics/metrics.go

```diff
@@ -0,0 +1,158 @@
+/* Copyright (c) 2023 Snowflake Inc. All rights reserved.
+
+   Licensed under the Apache License, Version 2.0 (the
+   "License"); you may not use this file except in compliance
+   with the License.  You may obtain a copy of the License at
+
+     http://www.apache.org/licenses/LICENSE-2.0
+
+   Unless required by applicable law or agreed to in writing,
+   software distributed under the License is distributed on an
+   "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY
+   KIND, either express or implied.  See the License for the
+   specific language governing permissions and limitations
+   under the License.
+*/
+// package metrics contains code for adding
+// metric instrumentations to Sansshell processes
+package metrics
+
+import (
+	"context"
+	"fmt"
+	"sync"
+
+	"github.com/pkg/errors"
+
+	"go.opentelemetry.io/otel/attribute"
+	"go.opentelemetry.io/otel/metric"
+	"go.opentelemetry.io/otel/metric/global"
+	"go.opentelemetry.io/otel/metric/instrument"
+)
+
+var (
+	// Singelton for instrumenting metrics
+	m *Metrics
+
+	errNotInitialized error = errors.New("metrics is not initialized. please call InitMetrics() before anything else")
+)
+
+type Metrics struct {
+	enabled bool
+	prefix  string
+	Meter   metric.Meter
+
+	// type: map[string]Int64Counter
+	Int64Counters sync.Map
+	// type: map[string]Int64Gauge
+	Int64Gauges sync.Map
+}
+
+type Option interface {
+	apply(*Metrics) error
+}
+
+type optionFunc func(*Metrics) error
+
+func (o optionFunc) apply(m *Metrics) error {
+	return o(m)
+}
+
+// addPrefix returns prefix + "_" + name if prefix is not empty
+// otherwise, it returns unaltered name
+func addPrefix(prefix, name string) string {
+	if prefix != "" {
+		name = fmt.Sprintf("%s_%s", m.prefix, name)
+	}
+	return name
+}
+
+// WithMetricNamePrefix stores the given prefix to the Metrics singleton
+func WithMetricNamePrefix(prefix string) Option {
+	return optionFunc(func(m *Metrics) error {
+		m.prefix = prefix
+		return nil
+	})
+}
+
+// InitMetrics initializes this package by creating the Metrics singleton
+// and applying options.
+// IMPORTANT: you need to call this before adding instrumentations
+func InitMetrics(opts ...Option) error {
```

Comment:
I'm not a big fan of doing this particular way of global singletons.

What do you think of instead copying the logr pattern of sticking this in a context and having top-level functions to create a context with it and to pull it out of the context?

```
func NewContext(context.Context, Metrics) context.Context
func FromContext(context.Context) (Metrics, bool)
```

Reply:
I wasn't sure between singleton and context-based, and thought singleton would be simpler.
But, I like the idea of copying the logr pattern because it makes the code more consistent.